Mobile/ShopLens/Droid/Activities/MainActivity.cs

```diff
@@ -103,59 +115,145 @@ protected override void OnCreate(Bundle savedInstanceState)
+        {
+            needUserAnswerId = ConfigurationManager.AppSettings["AnswerUtteranceId"];
+            askUserToRepeat = ConfigurationManager.AppSettings["AskToRepeat"];
+
+            ttsListener = new ShopLensUtteranceProgressListener(TtsStoppedSpeaking);
+
+            tts = new TextToSpeech(this, this);
+            tts.SetOnUtteranceProgressListener(ttsListener);
+
+            voiceRecognizer = new ShopLensSpeechRecognizer(OnVoiceRecognitionResults);
+
+            noTalkBack = true;
         }
 
-        // When the current voice recognition session stops.
-        public void OnResults(Bundle results)
+        private bool IsTutorialNeeded()
         {
-            var matches = results.GetStringArrayList(SpeechRecognizer.ResultsRecognition);
-            if (matches.Count > 0)
+            bool firstTime = prefs.GetBoolean("FirstTime", true);
+
+            if (firstTime)
             {
-                if (matches[0] == ConfigurationManager.AppSettings["CmdOpenCamera"])
-                {
-                    var intent = new Intent(this, typeof(CameraActivity));
-                    StartActivity(intent);
-                }
-                else if (matches[0] == ConfigurationManager.AppSettings["CmdOpenCart"])
-                {
-                    var intent = new Intent(this, typeof(ShoppingCartActivity));
-                    StartActivity(intent);
-                }
-                else if (matches[0] == ConfigurationManager.AppSettings["CmdOpenList"])
-                {
-                    var intent = new Intent(this, typeof(ShoppingListActivity));
-                    StartActivity(intent);
-                }
-                // For debugging purposes.
-                else
-                {
-                    voiceCommandButton.Text = matches[0];
-                }
+                prefs.Edit().PutBoolean("FirstTime", false).Commit();
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        private void RunUserTutorial()
+        {
+            var message = ConfigurationManager.AppSettings["InitialTutorialMsg"];
+            tts.Speak(message, QueueMode.Flush, null, needUserAnswerId);
+        }
+
+        private void ContinueUserTutorial()
+        {
+            var message = ConfigurationManager.AppSettings["FollowUpTutorialMsg"];
+            tts.Speak(message, QueueMode.Flush, null, needUserAnswerId);
+        }
+
+        // TTS Engine method called when TTS is initialized.
+        public void OnInit([GeneratedEnum] OperationResult status)
+        {
+            if (status == OperationResult.Error)
+            {
+                tts.SetLanguage(Locale.Default);
+            }
+            else if (status == OperationResult.Success)
+            {
+                tts.SetLanguage(Locale.Us);
+            }
+
+            if (tutorialNeeded)
+            {
+                RunUserTutorial();
```

Comment:
There's a way to move these commands out of OnInit, but it involves somehow finding out when OnInit is called and only then using tts.Speak. If you use it before OnInit is called, tts.Speak won't speak, it won't throw exceptions either, it will just be silent...

Reply:
This is the main reason why I didn't end up making a separate TTS helper class - we really need to write specific code for each activity in OnInit and it would look pretty ugly all in one class. That would make a separate TTS class dependant on our activity classes and it wouldn't be separate at all...